signature/plugin.go

```diff
@@ -99,81 +99,50 @@ func (s *pluginSigner) generateSignature(ctx context.Context, desc notation.Desc
 		return nil, fmt.Errorf("keyID in describeKey response %q does not match request %q", key.KeyID, s.keyID)
 	}
 
-	// Get algorithm associated to key.
-	alg := key.KeySpec.SignatureAlgorithm()
-	if alg == "" {
-		return nil, fmt.Errorf("keySpec %q for key %q is not supported", key.KeySpec, key.KeyID)
-	}
-
 	// Generate payload to be signed.
-	payload := packPayload(desc, opts)
-	if err := payload.Valid(); err != nil {
-		return nil, err
-	}
-
-	// Generate signing string.
-	token := jwtToken(alg.JWS(), payload)
-	payloadToSign, err := token.SigningString()
+	payload := notation.Payload{TargetArtifact: desc}
+	payloadBytes, err := json.Marshal(payload)
 	if err != nil {
-		return nil, fmt.Errorf("failed to marshal signing payload: %v", err)
+		return nil, fmt.Errorf("envelope payload can't be marshaled: %w", err)
 	}
 
-	// Execute plugin sign command.
-	req := &plugin.GenerateSignatureRequest{
-		ContractVersion: plugin.ContractVersion,
-		KeyID:           s.keyID,
-		KeySpec:         key.KeySpec,
-		Hash:            alg.Hash(),
-		Payload:         []byte(payloadToSign),
-		PluginConfig:    config,
-	}
-	out, err := s.runner.Run(ctx, req)
-	if err != nil {
-		return nil, fmt.Errorf("generate-signature command failed: %w", err)
-	}
-	resp, ok := out.(*plugin.GenerateSignatureResponse)
-	if !ok {
-		return nil, fmt.Errorf("plugin runner returned incorrect generate-signature response type '%T'", out)
+	// Create plugin signature provider
+	psp := pluginSigProvider{
+		runner:  s.runner,
+		ctx:     ctx,
+		keyID:   s.keyID,
+		keySpec: key.KeySpec,
 	}
-
-	// Check keyID is honored.
-	if s.keyID != resp.KeyID {
-		return nil, fmt.Errorf("keyID in generateSignature response %q does not match request %q", resp.KeyID, s.keyID)
+	signReq := signer.SignRequest{
+		Payload:             payloadBytes,
+		PayloadContentType:  signer.PayloadContentTypeV1,
+		SignatureProvider:   psp,
+		SigningTime:         time.Now(),
+		ExtendedSignedAttrs: nil,
+		SigningAgent:        "Notation/1.0.0",
 	}
-
-	// Check algorithm is supported.
-	jwsAlg := resp.SigningAlgorithm.JWS()
-	if jwsAlg == "" {
-		return nil, fmt.Errorf("signing algorithm %q in generateSignature response is not supported", resp.SigningAlgorithm)
+	if !opts.Expiry.IsZero() {
+		signReq.Expiry = opts.Expiry
 	}
 
-	// Check certificate chain is not empty.
-	if len(resp.CertificateChain) == 0 {
-		return nil, errors.New("generateSignature response has empty certificate chain")
+	// perform signing plugin signature provider
+	sigEnv, err := signer.NewSignatureEnvelope(signer.MediaTypeJWSJson)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal signing payload: %v", err)
 	}
```

Comment:
This seems unrelated to creating a new signature envelope with `NewSignatureEnvelope` , which only throws `UnsupportedSignatureFormatError`.